### PR TITLE
Gimbal v2 changes

### PR DESCRIPTION
**Pull request by julianoes**
This is a collection of changes/improvements on the initial gimbal v2 spec.
The changes mostly come out of the discussion with @olliw42 in #1419 and through experience after some implementation of the v2 protocol.

Basically, we found that the deconfliction/prioritization rules were not very clear and not covering more than a few specific cases. After some more iterations on how this could be done, @olliw42 and I gave up on such a design partly because we couldn't agree on the specifics and partly because it seemed to get very complex for not much benefit. We therefore went back to a similar approach as it was before with a configure message, although somewhat different. Instead of setting up the mode of the gimbal, we now set the sysid/compid of the component in control. The component can then use whichever messages and mode it needs. Additionally, there can be secondary input from a component for things like nudging, reposition, etc..

What we also determined was that for the most part all that's required is really just manual control without a unit, so basically just pan and tilt from -1 to 1, either in angle mode or angular rate mode. The gimbal can use whatever smoothness/speed makes sense, or is configured.

Note that the concept of gimbal manager and device stays mostly the same as before.

For more detail, check the individual commits.

Further todos:
- [x] Clean up flags
- [x] Check status and information messages
- [x] Update docs
- [ ] Figure out remaining questions in #1419.

### CODE REVIEW

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -6187,6 +6183,31 @@
       <field type="float" name="tilt_rate" units="rad/s">Tilt/pitch angular rate (positive: up, negative: down, NaN to be ignored).</field>
       <field type="float" name="pan_rate" units="rad/s">Pan/yaw angular rate (positive: to the right, negative: to the left, NaN to be ignored).</field>
     </message>
+    <message id="288" name="GIMBAL_MANAGER_SET_MANUAL_CONTROL">
+      <wip/>
+      <!-- This message is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+      <description>High level message to control a gimbal manually, so without units. The actual angles or angular rates will be produced by the gimbal manager based on settings. This message is to be sent to the gimbal manager (e.g. from a ground station). Angles and rates can be set to NaN according to use case.</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:6189]: "based on settings". By settings, do you mean "settings that are outside of the scope of MAVLink"? I.e. I think you're just saying that "the rates depend on gimbal hardware and configuration - so you can't set actual rates, just as a proportion of the total.

I guess also that the implementation might choose to not do this proportionally - but follow some rate curve based on the value.

File: message_definitions/v1.0/common.xml [hunk 2]
```diff
@@ -6187,6 +6183,31 @@
...
+      <field type="uint8_t" name="target_system">System ID</field>
+      <field type="uint8_t" name="target_component">Component ID</field>
+      <field type="uint32_t" name="flags" enum="GIMBAL_MANAGER_FLAGS">High level gimbal manager flags to use.</field>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:6192]: " to use" is redundant

File: message_definitions/v1.0/common.xml (hunk 1 above)
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:6189]: ", so without units" does not follow the "High level message to control a gimbal manually" logically. ie just because it is a high level message that does not mean that it has to be unitless. 

Perhaps "High level message to control a gimbal manually. The angles or angular rates are unitless; the actual rates will depend on internal gimbal manager settings/configuration. ...

File: message_definitions/v1.0/common.xml [hunk 3]
```diff
@@ -6187,6 +6183,31 @@
...
+      <field type="float" name="pan">Pan/yaw angle unitless (-1..1, positive: to the right, negative: to the left, NaN to be ignored).</field>
+      <field type="float" name="tilt_rate">Tilt/pitch angular rate unitless (-1..1, positive: up, negative: down, NaN to be ignored).</field>
+      <field type="float" name="pan_rate">Pan/yaw angular rate unitless (-1..1, positive: to the right, negative: to the left, NaN to be ignored).</field>
+    </message>
+    <message id="289" name="GIMBAL_DEVICE_SET_MANUAL_CONTROL">
+      <wip/>
+      <!-- This message is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+      <description>Low level message to control a gimbal manually, so without units. The actual angles or angular rates will be produced by the gimbal manager based on settings and by the gimbal device if available. This message is sent by the gimbal manager to the gimbal device. Angles and rates can be set to NaN according to use case. The message should not be sent at the same time as GIMBAL_DEVICE_SET_ATTITUDE, otherwise the two setpoints would conflict.</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:6202]: Same comment about "so without units" - that bit belongs in the next sentence.

File: message_definitions/v1.0/common.xml [hunk 4]
```diff
@@ -1732,6 +1714,16 @@
...
       </entry>
+      <entry value="1001" name="MAV_CMD_DO_GIMBAL_MANAGER_CONFIGURE" hasLocation="false" isDestination="false">
+        <wip/>
+        <!-- This message is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+        <description>Gimbal configuration to set which sysid/compid is in primary and secondary control.</description>
+        <param index="1" label="sysid primary control">Sysid currently in primary control (0: no one in control).</param>
+        <param index="2" label="compid primary control">Compid currently in primary control (0: no one in control).</param>
+        <param index="3" label="sysid secondary control">Sysid currently in secondary control (0: no one in control).</param>
+        <param index="4" label="compid secondary control">Compid currently in secondary control (0: no one in control).</param>
+        <param index="7" label="Gimbal device ID">Component ID of gimbal device to address (or 1-6 for non-MAVLink gimbal), 0 for all gimbal device components. (Send command multiple times for more than one but not all gimbals.)</param>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:1725]: "components. (Send command multiple times for more than one but not all gimbals.)" to either
- "components (send command multiple" or 
- "components. Send command multiple times"

File: message_definitions/v1.0/common.xml [hunk 5]
```diff
@@ -1732,6 +1714,16 @@
         <param index="5" label="Gimbal manager flags" enum="GIMBAL_MANAGER_FLAGS">Gimbal manager flags to use.</param>
         <param index="7" label="Gimbal device ID">Component ID of gimbal device to address (or 1-6 for non-MAVLink gimbal), 0 for all gimbal device components. (Send command multiple times for more than one but not all gimbals.)</param>
       </entry>
+      <entry value="1001" name="MAV_CMD_DO_GIMBAL_MANAGER_CONFIGURE" hasLocation="false" isDestination="false">
+        <wip/>
+        <!-- This message is work-in-progress and it can therefore change. It should NOT be used in stable production environments. -->
+        <description>Gimbal configuration to set which sysid/compid is in primary and secondary control.</description>
+        <param index="1" label="sysid primary control">Sysid currently in primary control (0: no one in control).</param>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml:1721]: For all of these do you mean "currently in" or "Sysid to take primary control"

Comment (julianoes) [reply message_definitions/v1.0/common.xml:1721]: It means that this component is in control (not to take). It means that messages arriving from that sysid/compid will be used/applied.

**Comment by julianoes** [comment]
Thanks for the review @hamishwillee. I'll do a pass later.

File: message_definitions/v1.0/common.xml (hunk 5 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml:1721]: @julianoes English doesn't make sense (to me). As I understand it the  ownership of the gimbal and who is the primary and secondary is managed by the gimbal manager. So the manager broadcasts the current controller(s) using GIMBAL_MANAGER_STATUS. Another system can set itself as the "current controller" using *this* message MAV_CMD_DO_GIMBAL_MANAGER_CONFIGURE.

The point is that this is a setter so you probably shouldn't use "currently" because it makes it sound like the message is a getter - getting the current controller id. Also it is redundant.

Suggested changes:
- "Sysid currently in primary control (0: no one in control)." TO "Sysid for primary control (0: no one in control)"
- Same for the other fields.
- In the description, it is worth saying here something like "Check current controllers in GIMBAL_MANAGER_STATUS before arbitrarily seizing control. For more info see ....)

Comment (julianoes) [reply message_definitions/v1.0/common.xml:1721]: Oh, sorry, you're right, thx!

File: message_definitions/v1.0/common.xml (hunk 1 above)
Comment (julianoes) [reply message_definitions/v1.0/common.xml:6189]: Yes, it would bee based on settings outside of MAVLink, e.g. params.

Comment (julianoes) [reply message_definitions/v1.0/common.xml:6189]: Thx, yes!

**Comment by julianoes** [comment]
Thanks @hamishwillee. I've added one commit trying to address your review comments.

**Comment by hamishwillee** [comment]
@julianoes Much better thanks. I did a minor subedit for consistency. 

The only concern I have may well be resolved by the other docs - still don't understand the model for primary and secondary control. But aside from that this is good to go (feel free to self-merge)

**Comment by julianoes** [comment]
> still don't understand the model for primary and secondary control.

That's up to the implementation and not really specified by the protocol. For some use cases there will only be a primary, and for others there might be a primary e.g. for the general angle and a secondary for slight nudging, and others might just add the two sources together.

**Comment by hamishwillee** [comment]
> That's up to the implementation and not really specified by the protocol. For some use cases there will only be a primary, and for others there might be a primary e.g. for the general angle and a secondary for slight nudging, and others might just add the two sources together.

Not just that - see my comments in the docs PR. Specifically what happens if multiple things try to set these and the interactions. Can I set myself as the secondary if something else has set itself as primary etc. Who owns what.

**Comment by julianoes** [comment]
> Can I set myself as the secondary if something else has set itself as primary etc. Who owns what.

Yes, anyone can change it as they want. I think that's something that we should explain in the FAQ part of the docs.

**Comment by julianoes** [comment]
I have come to the conclusion - at least for now - that we should only have the manual control message from ground station/companion/other clients to the gimbal manager, however, not from the gimbal manager to the gimbal device.

Here is my reasoning:
- The gimbal device does not know how to mix an angle setpoint and a manual control setpoint (e.g. nudging). This is rather something that can/should be done by the gimbal manager.
- The gimbal device would get confused if there are attitude and manual control messages arriving at the same time. It's not specified what should happen then. If possible, I would like to prevent that case by protocol design.
- One of advantages of manual control is that the pan/tilt speed can get adjusted to a camera zoom level so that panning is nice and slow when zoomed in. This would happen in the gimbal manager because a gimbal with integrated camera is typically the gimbal manager (and not device).

What's not so clear?
- The gimbal device knows its limits/settings like angular rate and smoothness best. It does communicate max angles and max angular velocities using GIMBAL_DEVICE_INFORMATION which allows the gimbal manager to translate manual control setpoints to angular velocities, or angles, but it doesn't seem necessarily seem optimal.

I'm going to merge this PR now and work on the implementation. Once I find how it does or does not work, I'll come back to the protocol which is still marked WIP.
